### ISSUE TITLE
are these tensor shape correct ?

### ISSUE THREAD

**Issue by vicentowang** [eladrich/pixel2style2pixel#218]
![image](https://user-images.githubusercontent.com/49265512/140875943-5908a0a7-8be7-4cc1-bd15-729ef52f3d96.png)
latent code shape is kind of different with the paper. why y_hat is batch size of 4 ?

**Comment by yuval-alaluf** [comment]
If you are using a generator with a resolution of 512 then the latent code should have 16 entries so seems fine. In our paper, we used an output resolution of 1024 which uses 18 entries

**Comment by vicentowang** [comment]
@yuval-alaluf  y_hat batch size is 4 ? means one image input, generate 4 image?

**Comment by yuval-alaluf** [comment]
The batch size of 4 means the number of input images. The number of output images is equal to the number of inputs - you get one output for each input

**Comment by vicentowang** [comment]
@yuval-alaluf  x tensor batch size is 1 , after forward pass ,got y_hat batch size of 4 ? am I any config wrong?

**Comment by MKFMIKU** [comment]
Facing the same issue.

**Comment by vicentowang** [comment]
@MKFMIKU  @yuval-alaluf  I add a conv layer to solved problem.

class GradualStyleBlock(Module):
    def __init__(self, in_c, out_c, spatial):
        super(GradualStyleBlock, self).__init__()
        self.out_c = out_c
        self.spatial = spatial
        num_pools = int(np.log2(spatial))
        modules = []
        modules += [Conv2d(in_c, out_c, kernel_size=3, stride=2, padding=1),
                    nn.LeakyReLU()]
        for i in range(num_pools): # here
            modules += [
                Conv2d(out_c, out_c, kernel_size=3, stride=2, padding=1),
                nn.LeakyReLU()
            ]
        self.convs = nn.Sequential(*modules)
        self.linear = EqualLinear(out_c, out_c, lr_mul=1)

    def forward(self, x):
        x = self.convs(x)
        x = x.view(-1, self.out_c)
        x = self.linear(x)
        return x

**Comment by yuval-alaluf** [comment]
You shouldn't need to change any of the architecture, but glad to see this solved your issue.

**Comment by vicentowang** [comment]
@yuval-alaluf  the problem is caused by the convolution padding with different size of feature map.

**Comment by vicentowang** [comment]
@yuval-alaluf  can you fix the shape problem mentioned above, not just me face the issue. and it is exist. I change the code,but cannot get the right resule of image generation.

**Comment by yuval-alaluf** [comment]
> the problem is caused by the convolution padding with different size of feature map. since you use the view() tensor operation.

Not sure i follow. What do you mean by "different size of feature map"? If you have not made any changes to the inner workings of the code, there should be no issue. Can you please provide more details to your data resolution, stylegan output size, etc.

**Comment by vicentowang** [comment]
@yuval-alaluf    stylegan output size is 512x512 , 
![](https://user-images.githubusercontent.com/49265512/140875943-5908a0a7-8be7-4cc1-bd15-729ef52f3d96.png)
input x size is [1, 3, 512,512], ground truth y size is [1, 3, 512,512]
output y_hat shape should be [1, 3, 512,512], but I got  [4, 3, 512,512] , problem is that loss(y,y_hat) cannot be caculated,since shape is different.

**Comment by yuval-alaluf** [comment]
Ok. This seems to be caused because you are using a batch size of 1. Does this occur if you run with a batch size of 2?

**Comment by vicentowang** [comment]
@yuval-alaluf  I  haven't try that yet. since my GPU memory is not enough, so I use batch size of 1

**Comment by yuval-alaluf** [comment]
I believe this is what is causing the issue. Working with a batch size of 1 results in unwanted changes to the tensor dimensions. Some changes will probably be needed to support a batch size of 1.

**Comment by vicentowang** [comment]
@yuval-alaluf  Thanks for your concern.

**Comment by vicentowang** [comment]
@yuval-alaluf 
not really.  batch size set to 2, got y_hat batch size of 8.
![image](https://user-images.githubusercontent.com/49265512/144779906-9c5c0c62-c46d-4018-9bc1-aa43e025f7ca.png)

**Comment by yuval-alaluf** [comment]
@vicentowang , I have just run the code and there is no problem with the code. believe you have either changed something in the code or changed the transforms.

Take a look at the original transforms: 
https://github.com/eladrich/pixel2style2pixel/blob/7a511c687bf2a8a64ba6a47150b37c7108329a6a/configs/transforms_config.py#L21-L37

You can see here that the images are resized to 256. Your images are of size `512` which tells me you have changed something. Please check that your code, transforms, and data are all correct. 

As explained in the paper, regarding the face pool, since our inputs are of size 256, we resized the outputs to 256 so that we can compute the loss. This allows us to use lower resolution inputs during training to speed up training. At inference, however, you can still get the full 1024x1024 outputs.

**Comment by vicentowang** [comment]
here is the problem. self.encoder generate 4 times of code. it's batch size should be the same as x 
![image](https://user-images.githubusercontent.com/49265512/144796625-4a214018-6571-42cc-b4be-a3e37229b34d.png)

**Comment by yuval-alaluf** [comment]
Did you change the transforms? Based on the size of x (which is 512x512), you did change something.

**Comment by vicentowang** [comment]
![image](https://user-images.githubusercontent.com/49265512/144797457-a05b0d38-bbb4-4c54-889b-97da5ceb8571.png)
no scale change since my image input is all 512x512

**Comment by yuval-alaluf** [comment]
So you did change the code. Put back the rescaling to 256 and see if it fixes your problem. 
If you change the code and something doesn't work, it most likely means the change you made was the source of the problem.

**Comment by vicentowang** [comment]
@ I feel confused that changes to the EncodeTransforms affect the network structure output. but I cannot add  transforms.Resize((256, 256)), because my input image is not png or jpg etc. have no api-resize for EXR image.

**Comment by yuval-alaluf** [comment]
If you change the input size, it will change the output size since we've working with convolutions. 
If you can't find a way to resize the input image, you will probably need to add another downsampling layer in the network.

**Comment by vicentowang** [comment]
here , x is the network input ,which i ensure it shape is 512x512
![image](https://user-images.githubusercontent.com/49265512/144799349-b3d5b1cd-e3fd-4907-be40-411988964660.png)

**Comment by vicentowang** [comment]
I set image size to 256, and now its working, codes = self.encoder(x)  codes and x batch size is the same as expected.
![image](https://user-images.githubusercontent.com/49265512/144800152-5d4a095e-e8bd-4a52-b783-25585a446023.png)

**Comment by yuval-alaluf** [comment]
I dont see a reason to train on an input resolution of 512. We showed great results even when using inputs of 256. 
However, if you insist on training with 512, you need to add another downsampling layer in the GradualStyleBlock

**Comment by vicentowang** [comment]
i got it , and another problem , I trained a stylegan2-ada model(.pkl model file) which input is W space, and how to made it to receive input of W+ space.

**Comment by yuval-alaluf** [comment]
pSp already handles working with W+. There is no change needed

**Comment by vicentowang** [comment]
@ which repo code of stylegan your paper use ?
![Uploading image.png…]()
my pretrained stylegan model file seem to diff from your model.

**Comment by yuval-alaluf** [comment]
All the details are in the readme. It is incredibly detailed.

**Comment by vicentowang** [comment]
.pkl loaded error in your code stylegan structure.
my pretrained stylegan2-ada model may be kind of different to yours, so I have to build stylegan-ada structure in W+ space.
